### DISCUSSION TITLE
Laravel Reverb in Production

### DISCUSSION THREAD

**Discussion by amypo-tech** [laravel/framework#50675] [General]
I have created a sample laravel 11 project and in which I have configured all the requirements for the laravel reverb.

In local machine/setup reverb is working fine. when I deployed in production server for testing I am getting the bellow error.
<img width="1440" alt="Screenshot 2024-03-20 at 11 15 08 PM" src="https://github.com/laravel/framework/assets/74060536/98265f86-d811-4eb4-b3ff-39dc6c668435">

- Defined SSL cert and ket path correctly
- port 8080 is not blocked which can be accessed from any machine
- tried with and without https for reverb hosting server
- tried with localhost as reverb host

None of the above helped me. Kindly help me on this.

**.env**

```
APP_NAME=Laravel
APP_ENV=local
APP_KEY=base64:8aOIIcWglqC0snCZJgA33uCG68BQ2riXec7qLR2M8pg=
APP_DEBUG=true
APP_TIMEZONE=UTC
APP_URL=https://reverb.amypo.com

APP_LOCALE=en
APP_FALLBACK_LOCALE=en
APP_FAKER_LOCALE=en_US

APP_MAINTENANCE_DRIVER=file
APP_MAINTENANCE_STORE=database

BCRYPT_ROUNDS=12

LOG_CHANNEL=stack
LOG_STACK=single
LOG_DEPRECATIONS_CHANNEL=null
LOG_LEVEL=debug

DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=reverb
DB_USERNAME=root
DB_PASSWORD='0birj@*RM}'

SESSION_DRIVER=database
SESSION_LIFETIME=120
SESSION_ENCRYPT=false
SESSION_PATH=/
SESSION_DOMAIN=null

BROADCAST_CONNECTION=reverb
FILESYSTEM_DISK=local
QUEUE_CONNECTION=database

CACHE_STORE=database
CACHE_PREFIX=

MEMCACHED_HOST=127.0.0.1

REDIS_CLIENT=phpredis
REDIS_HOST=127.0.0.1
REDIS_PASSWORD=null
REDIS_PORT=6379

MAIL_MAILER=log
MAIL_HOST=127.0.0.1
MAIL_PORT=2525
MAIL_USERNAME=null
MAIL_PASSWORD=null
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS="hello@example.com"
MAIL_FROM_NAME="${APP_NAME}"

AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
AWS_DEFAULT_REGION=us-east-1
AWS_BUCKET=
AWS_USE_PATH_STYLE_ENDPOINT=false

VITE_APP_NAME="${APP_NAME}"

REVERB_APP_ID=864321
REVERB_APP_KEY=xypigyotzowvifsmqzuw
REVERB_APP_SECRET=owz65aaetqlxgjqvd8s3
REVERB_HOST="websockets.amypo.com"
REVERB_PORT=8080
REVERB_SCHEME=https

VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
VITE_REVERB_HOST="${REVERB_HOST}"
VITE_REVERB_PORT="${REVERB_PORT}"
VITE_REVERB_SCHEME="${REVERB_SCHEME}"
```

**config/reverb.php**

 ```
   'servers' => [

        'reverb' => [
            'host' => env('REVERB_SERVER_HOST', '0.0.0.0'),
            'port' => env('REVERB_SERVER_PORT', 8080),
            'hostname' => env('REVERB_HOST'),
            'options' => [
                 'tls' => [
                'local_cert' => '/etc/letsencrypt/live/websockets.amypo.com/fullchain.pem',
                'local_pk' => '/etc/letsencrypt/live/websockets.amypo.com/privkey.pem',
            ],
            ],
            'max_request_size' => env('REVERB_MAX_REQUEST_SIZE', 10_000),
            'scaling' => [
                'enabled' => env('REVERB_SCALING_ENABLED', false),
                'channel' => env('REVERB_SCALING_CHANNEL', 'reverb'),
            ],
            'pulse_ingest_interval' => env('REVERB_PULSE_INGEST_INTERVAL', 15),
        ],

    ],


    'apps' => [

        'provider' => 'config',

        'apps' => [
            [
                'key' => env('REVERB_APP_KEY'),
                'secret' => env('REVERB_APP_SECRET'),
                'app_id' => env('REVERB_APP_ID'),
                'options' => [
                    'host' => env('REVERB_HOST'),
                    'port' => env('REVERB_PORT', 443),
                    'scheme' => env('REVERB_SCHEME', 'https'),
                    'useTLS' => env('REVERB_SCHEME', 'https') === 'https',
                ],
                'allowed_origins' => ['*'],
                'ping_interval' => env('REVERB_APP_PING_INTERVAL', 60),
                'max_message_size' => env('REVERB_APP_MAX_MESSAGE_SIZE', 10_000),
            ],
        ],

    ],
```

**Comment by ITApexGroup** [comment]
In Env
..
```
REVERB_APP_ID=864321
REVERB_APP_KEY=xypigyotzowvifsmqzuw
REVERB_APP_SECRET=owz65aaetqlxgjqvd8s3
REVERB_HOST="example.com"
REVERB_PORT=443
REVERB_SCHEME=https
```

In config/reverb.php
..
```
'options' => [
            'tls' => [
           ],
],
```

In the project nginx server config
..
```
location /app {
        proxy_http_version 1.1;
        proxy_set_header Host $http_host;
        proxy_set_header Scheme $scheme;
        proxy_set_header SERVER_PORT $server_port;
        proxy_set_header REMOTE_ADDR $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";

        proxy_pass http://0.0.0.0:8080;
}
```

It is working for me.

**Reply by amypo-tech** [reply]
Thank you for responding and spending your valuable time.

I have updated the `.env`

But I am using `Apache server `should I use `Nginx` along with apache and run as reverse proxy server.

Can you tell me the setup. That will help me a lot.

Thanks in advance.

**Reply by matheusbento** [reply]
It worked for me, thanks.

**Reply by Freddywhest** [reply]
For Apache server try the below

```
<IfModule mod_proxy.c>
    RewriteEngine On
    RewriteCond %{REQUEST_URI} ^/app
    RewriteRule ^(.*)$ http://0.0.0.0:8080/$1 [P]

    <Proxy http://0.0.0.0:8080>
        ProxyPreserveHost On
        ProxyPassReverse /
        ProxyPassReverseCookiePath / /app
    </Proxy>
</IfModule>
```

**Reply by webmid** [reply]
Where should we put this code? In vhost.conf?
I tried but still I receive connection failed in client. this is my env var:

`
BROADCAST_CONNECTION=reverb
REVERB_APP_ID=xxxx
REVERB_APP_KEY=xxxx
REVERB_APP_SECRET=xxxx
REVERB_HOST="mydomain.de"
REVERB_PORT=443
REVERB_SCHEME=https
REVERB_SERVER_PORT=8080


VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
VITE_REVERB_HOST="${REVERB_HOST}"
VITE_REVERB_PORT="${REVERB_PORT}"
VITE_REVERB_SCHEME="${REVERB_SCHEME}"
`
I'm using Apache and ssl

Thanks

**Reply by Ivan3008** [reply]
thank you! it works

**Reply by repositoriomaster** [reply]
It is working to me.
add location /app {...
TKU!

**Reply by Kenneth-Inkum** [reply]
You're a lifesaver. Such simple solution.

**Reply by eslam-eng** [reply]
what is path of file  that i should put this code : 
`<IfModule mod_proxy.c>
    RewriteEngine On
    RewriteCond %{REQUEST_URI} ^/app
    RewriteRule ^(.*)$ http://0.0.0.0:8080/$1 [P]

    <Proxy http://0.0.0.0:8080>
        ProxyPreserveHost On
        ProxyPassReverse /
        ProxyPassReverseCookiePath / /app
    </Proxy>
</IfModule>
`
@Freddywhest

**Comment by amypo-tech** [comment]
I have the same type of issue. nothing was helpful.

implementing web-sockets in production is like finding water in mars.

All the possible ways i have tried.

1. all the possible .env configurations
2. nginx reverse proxy configuration
3. config/reverb.php config changes with and without certs and different ports

if some one gives a proper setup or guide from start to end would help for production.


in local it took only few minutes to implement. this is the same thing happened before launch of reverb. waiting nearly 6 months to implement in production

**Reply by dominik-eller** [reply]
I am having the exact same issue even with using Laravel Forge which should be working out of the box because it should configure everything you need.
But I am stuck and have no idea how to solve that.

There seems to be non working example on Laravel Reverb using Laravel Forge and the documentation is not very detailed on that point.

Would be great to get some help and some working examples.


**EDIT:**
I deactivated and activated Laravel Reverb in die Laravel Forge Application Panel and restarted nginx and it is working now. Maybe the nginx config for the reverse proxy was not set correctly or nginx was not reloaded automatically.

The working configuration in my .env. file:

    REVERB_APP_ID=***
    REVERB_APP_KEY=***
    REVERB_APP_SECRET=***
    REVERB_HOST=127.0.0.1
    REVERB_PORT=8080
    REVERB_SCHEME=http

    VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
    VITE_REVERB_HOST=ws.domain.tld
    VITE_REVERB_PORT=443
    VITE_REVERB_SCHEME=https

**Reply by vpuentem** [reply]
+1

**Reply by Jhamnerx** [reply]
+1

**Reply by dnamxtvl** [reply]
+1

**Reply by hiccup711** [reply]
thank you very much, it helps

**Reply by weristwiegott** [reply]
> I am having the exact same issue even with using Laravel Forge which should be working out of the box because it should configure everything you need. But I am stuck and have no idea how to solve that.
> 
> There seems to be non working example on Laravel Reverb using Laravel Forge and the documentation is not very detailed on that point.
> 
> Would be great to get some help and some working examples.
> 
> **EDIT:** I deactivated and activated Laravel Reverb in die Laravel Forge Application Panel and restarted nginx and it is working now. Maybe the nginx config for the reverse proxy was not set correctly or nginx was not reloaded automatically.
> 
> The working configuration in my .env. file:
> 
> ```
> REVERB_APP_ID=***
> REVERB_APP_KEY=***
> REVERB_APP_SECRET=***
> REVERB_HOST=127.0.0.1
> REVERB_PORT=8080
> REVERB_SCHEME=http
> 
> VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
> VITE_REVERB_HOST=ws.domain.tld
> VITE_REVERB_PORT=443
> VITE_REVERB_SCHEME=https
> ```

THATS THE SOLUTION!!!

**Comment by thecodeholic** [comment]
I came across to this issue when I needed to setup reverb on VPS for myself.
After about 1 day digging into it I made it working on my machine and decide to come back to this and share it.
Here is what makes if working in my case.

Add the following server tag in your nginx configuration file.

```
server {
  listen 8081 ssl;
  listen [::]:8081 ssl;
  server_name yourserver.name;
  {{ssl_certificate_key}}
  {{ssl_certificate}}

    location / {
        proxy_http_version 1.1;
        proxy_set_header Host $http_host;
        proxy_set_header Scheme $scheme;
        proxy_set_header SERVER_PORT $server_port;
        proxy_set_header REMOTE_ADDR $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";

        proxy_pass http://127.0.0.1:8082;
    }
}
```

This is my `.env`

```
REVERB_APP_ID=*****
REVERB_APP_KEY=*****
REVERB_APP_SECRET=*****
REVERB_HOST="yourserver.name"
REVERB_PORT=8081
REVERB_SERVER_PORT=8082
REVERB_SCHEME=https
```

In my case port 8080 was used by another process and I started reverb on port 8082 and I am listening on nginx at port 8081 which passes the socket connection to reverb.

## [Update]
Also, Make sure that incoming port `8080` or `8081` (whichever you decided to connect to) is opened.

Check if it is opened

```
sudo ufw status verbose
```

Open port 8080

```
sudo ufw allow 8080
```

**Reply by mjohnson10111** [reply]
this is what fixed it for me, opening the port

**Comment by krzysztofengineer** [comment]
Here's my working configuration for Laravel Forge:

```REVERB_APP_ID=677755
REVERB_APP_KEY=APP-KEY
REVERB_APP_SECRET=APP-SECRET
REVERB_HOST=127.0.0.1
REVERB_PORT=8080
REVERB_SCHEME=http

VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
VITE_REVERB_HOST=ws.DOMAIN.LTD
VITE_REVERB_PORT=443
VITE_REVERB_SCHEME=https

**Reply by krzysztofengineer** [reply]
maybe not related to this thread, but if someone has a similar issue: if your websockets connection is working but you're still not broadcasting events from the server, make sure your queue is running (or change the QUEUE_CONNECTION to sync). It's not clear that ShouldBroadcast interface uses queues under the hood

**Comment by mjohnson10111** [comment]
i am running forge - oddly, i didnt have to edit the nginx files

##env
REVERB_APP_ID=XXXX
REVERB_APP_KEY=XXXX
REVERB_APP_SECRET=XXXX
REVERB_HOST=127.0.0.1
REVERB_PORT=8080
REVERB_SCHEME=http

VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
VITE_REVERB_HOST=ws.XXX.com
VITE_REVERB_PORT=443
VITE_REVERB_SCHEME=https

##CHECK PORTS ARE OPEN
sudo ufw status verbose

##OPEN THEM IF NOT
sudo ufw allow 8080

**Comment by amypo-tech** [comment]
In my case everything seems to be fine in the production server. 

running Nginx in port 9000
```
default.conf

server {
    listen 80 default_server;
    listen [::]:80 default_server;

    root /var/www/html;
    index index.html index.htm index.nginx-debian.html;

    server_name websockets.amypo.com;

    ssl_certificate_key /etc/letsencrypt/live/websockets.amypo.com/privkey.pem;
    ssl_certificate /etc/letsencrypt/live/websockets.amypo.com/fullchain.pem;


    location / {
        try_files $uri $uri/ =404;
    }

    # Reverse proxy configuration for /app
    location /app {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://127.0.0.1:9000; 
      
    }

    location ~ /\.ht {
        deny all;
    }
}

```

.env
```
APP_NAME=Laravel
APP_ENV=local
APP_KEY=base64:8aOIIcWglqC0snCZJgA33uCG68BQ2riXec7qLR2M8pg=
APP_DEBUG=true
APP_TIMEZONE=UTC
APP_URL=https://reverb.amypo.com

APP_LOCALE=en
APP_FALLBACK_LOCALE=en
APP_FAKER_LOCALE=en_US

APP_MAINTENANCE_DRIVER=file
APP_MAINTENANCE_STORE=database

BCRYPT_ROUNDS=12

LOG_CHANNEL=stack
LOG_STACK=single
LOG_DEPRECATIONS_CHANNEL=null
LOG_LEVEL=debug

DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=reverb
DB_USERNAME=root
DB_PASSWORD='XXXXXX'

SESSION_DRIVER=database
SESSION_LIFETIME=120
SESSION_ENCRYPT=false
SESSION_PATH=/
SESSION_DOMAIN=null

BROADCAST_CONNECTION=reverb
FILESYSTEM_DISK=local
QUEUE_CONNECTION=database

CACHE_STORE=database
CACHE_PREFIX=

MEMCACHED_HOST=127.0.0.1

REDIS_CLIENT=phpredis
REDIS_HOST=127.0.0.1
REDIS_PASSWORD=null
REDIS_PORT=6379

MAIL_MAILER=log
MAIL_HOST=127.0.0.1
MAIL_PORT=2525
MAIL_USERNAME=null
MAIL_PASSWORD=null
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS="hello@example.com"
MAIL_FROM_NAME="${APP_NAME}"

AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
AWS_DEFAULT_REGION=us-east-1
AWS_BUCKET=
AWS_USE_PATH_STYLE_ENDPOINT=false

VITE_APP_NAME="${APP_NAME}"

REVERB_APP_ID=864321
REVERB_APP_KEY=xypigyotzowvifsmqzuw
REVERB_APP_SECRET=owz65aaetqlxgjqvd8s3
REVERB_HOST=127.0.0.1
REVERB_PORT=9000
REVERB_SCHEME=http

VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
VITE_REVERB_HOST=websockets.amypo.com
VITE_REVERB_PORT=443
VITE_REVERB_SCHEME=https
VITE_REVERB_SCHEME="${REVERB_SCHEME}"
REVERB_SCALING_ENABLED=true
```
revert.php
```
'servers' => [

        'reverb' => [
            'host' => env('REVERB_SERVER_HOST', '127.0.0.1'),
            'port' => env('REVERB_SERVER_PORT', 9000),
            'hostname' => env('REVERB_HOST'),
            'options' => [
                 'tls' => [
            ],
            ],
            'max_request_size' => env('REVERB_MAX_REQUEST_SIZE', 10_000),
            'scaling' => [
                'enabled' => env('REVERB_SCALING_ENABLED', false),
                'channel' => env('REVERB_SCALING_CHANNEL', 'reverb'),
            ],
            'pulse_ingest_interval' => env('REVERB_PULSE_INGEST_INTERVAL', 15),
        ],

    ],
```

```
root@localhost:/var/www/reverb.amypo.com/reverb_new# php artisan reverb:start

   INFO  Starting server on 0.0.0.0:9000 (127.0.0.1).  
```

```
root@localhost:/var/www/reverb.amypo.com/reverb_new# php artisan queue:listen

   INFO  Processing jobs from the [default] queue.  

  2024-04-07 06:38:48 App\Events\Reverbevent ......................... RUNNING
  2024-04-07 06:38:48 App\Events\Reverbevent .................... 64.98ms DONE
  2024-04-07 06:38:52 App\Events\Reverbevent ......................... RUNNING
  2024-04-07 06:38:52 App\Events\Reverbevent .................... 46.32ms DONE
  2024-04-07 06:40:11 App\Events\Reverbevent ......................... RUNNING
  2024-04-07 06:40:11 App\Events\Reverbevent .................... 81.67ms DONE
```

still getting the same error
`WebSocket connection to 'wss://websockets.amypo.com/app/xypigyotzowvifsmqzuw?protocol=7&client=js&version=8.4.0-rc2&flash=false' failed: `

Laravel Log:
```
[2024-04-07 06:30:37] local.ERROR: Pusher error: cURL error 7: Failed to connect to 127.0.0.1 port 9000 after 0 ms: Connection refused (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://127.0.0.1:9000/apps/864321/events?auth_key=xypigyotzowvifsmqzuw&auth_timestamp=1712471436&auth_version=1.0&body_md5=ba512609333df91c096d8893a3cf88a1&auth_signature=ae33b603c3286f00a95bc9bc96bab09e3f7f001390ee3aa7e3f63c4df16b3052. {"exception":"[object] (Illuminate\\Broadcasting\\BroadcastException(code: 0): Pusher error: cURL error 7: Failed to connect to 127.0.0.1 port 9000 after 0 ms: Connection refused (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://127.0.0.1:9000/apps/864321/events?auth_key=xypigyotzowvifsmqzuw&auth_timestamp=1712471436&auth_version=1.0&body_md5=ba512609333df91c096d8893a3cf88a1&auth_signature=ae33b603c3286f00a95bc9bc96bab09e3f7f001390ee3aa7e3f63c4df16b3052. at /var/www/reverb.amypo.com/reverb_new/vendor/laravel/framework/src/Illuminate/Broadcasting/Broadcasters/PusherBroadcaster.php:164)
[stacktrace]
```

**Reply by hrant1020** [reply]
@amypo-tech  Your `REVERB_SERVER_HOST` is equal to `REVERB_HOST`

[Look laravel reverb doc](https://laravel.com/docs/11.x/reverb#:~:text=configuration%20file.-,The,.,-The%20former%20specify)

**Comment by gamboitaygb** [comment]
Hello, let me give you my point,I encountered a similar issue, my localhost was working perfectly but in my traning server (VSP with WHM) it didn't work.
Since I am not using NGINX so the solution was over apache, my head was about to blow, cuz I also  using cloudflare too so, I looked there, but the problem was apache. so, Let me share my solution.

Firstly, ensure that the proxy module is enabled:


`apachectl -M`

You should see something like this:

```
proxy_module (shared)
proxy_fcgi_module (shared)
proxy_http_module (shared)
proxy_wstunnel_module (shared)
```
If the module isn't installed, you'll need to install it and load it into Apache (though this wasn't my case). You may also need to uncomment something in the httpd.conf file.

After that, insert this code into your virtual host or use an include file:


```
<IfModule mod_proxy.c>
    <IfModule mod_proxy_wstunnel.c>
        RewriteEngine On
        RewriteCond %{HTTP:Upgrade} =websocket [NC]
        RewriteCond %{HTTP:Connection} upgrade$ [NC]
        RewriteRule ^/app(.*)$ ws://127.0.0.1:8080/app$1 [P,L]

        ProxyPass /app ws://127.0.0.1:8080/app
        ProxyPassReverse /app ws://127.0.0.1:8080/app
    </IfModule>

    RewriteCond %{REQUEST_URI} ^/app [NC]
    RewriteRule ^/app(.*)$ http://127.0.0.1:8080/app$1 [P,L]

    ProxyPass /app http://127.0.0.1:8080/app
    ProxyPassReverse /app http://127.0.0.1:8080/app
</IfModule>
```
This solved my problem.

**Reply by felipeyepes29** [reply]
I have the same problem, I am using a godaddy VHM and I have not been able to get it to work, could you give more details on how you configured it?

**Reply by MuscledGeek** [reply]
This help me partially, I had to put these lines in virtual hosts:

```
ProxyPreserveHost on
ProxyPass "/app" "ws://0.0.0.0:8080/app"
ProxyPassReverse "/app" "ws://0.0.0.0:8080/app"
```

Just right before letsencrypt certs 💯

**Reply by realdealride** [reply]
Thanks. I'm success.

**Comment by meloDramatic007** [comment]
This is how I solved reverb integration problem on production server. I use docker+ssl enabled . I worte this medium article ,you can find the configuration here https://medium.com/@saddanfah/how-to-use-laravel-reverb-in-production-server-d89e9670b76a

**Comment by iandk** [comment]
In case anyone is having issues getting it to work with Caddy

```
# Caddyfile
yourdomain.tld {
    root * /home/web/public/public
    file_server
    encode gzip
    php_fastcgi unix//run/php/php8.2-web-fpm.sock {
        split .php
        index index.php
    }
    header {
        Strict-Transport-Security max-age=31536000
        Permissions-Policy interest-cohort=()
        X-Content-Type-Options nosniff
        X-Frame-Options SAMEORIGIN
        Referrer-Policy no-referrer
        X-XSS-Protection "1; mode=block"
        X-Permitted-Cross-Domain-Policies none
        X-Robots-Tag "noindex, nofollow"
        -X-Powered-By
    }
    handle_path /.well-known/acme-challenge/* {
        try_files {path} {path}/
        file_server
    }
    handle_path /.well-known/pki-validation/* {
        try_files {path} {path}/
        file_server
    }
    redir /.well-known/* /index.php{uri} 301

}


ws.yourdomain.tld {
    reverse_proxy 127.0.0.1:8080
}
```

```
# .env
VITE_APP_NAME="${APP_NAME}"

REVERB_APP_ID= XXXXXXXXX
REVERB_APP_KEY=XXXXXXXXX
REVERB_APP_SECRET= XXXXXXXXX
REVERB_HOST="ws.yourdomain.tld"
REVERB_PORT=443
REVERB_SCHEME=https

VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
VITE_REVERB_HOST="${REVERB_HOST}"
VITE_REVERB_PORT="${REVERB_PORT}"
VITE_REVERB_SCHEME="${REVERB_SCHEME}"
```

**Comment by iamvinny** [comment]
It took me a while but I managed to get it working, here are some very important things that can prevent Reverb from working on a production server:

1. The port 8080 *(or any other port you want to use)* must be open an accessible, you can do this by configuring your firewall when creating the server at your hosting provider. To verify if it is indeed open, you must start the reverb server with `php artisan reverb:start --debug`, then head over to https://www.yougetsignal.com/tools/open-ports/ and enter the server IP address and port you want to check.

2. You must ensure that the `.env` variables related to Reverb are correctly set, specially `BROADCAST_CONNECTION` which must be always set to `reverb`, it is default set to `log`. And also, ensure that `REVERB_HOST` is set to your domain name or server IP.

3. After you make any changes to the `.env` file, you must rebuild your node packages with `npm run build`.

**Comment by jihad28** [comment]
## Here is my solution for Secure Application (SSL)
### System info:
apache2
ssl certificate (letsencrypt)

  

### First update the `.env` file like this

```
REVERB_HOST=0.0.0.0
REVERB_PORT=8080
REVERB_SCHEME=http

VITE_REVERB_HOST="yourdomain.com" // Replace it with your domain
VITE_REVERB_PORT=443
VITE_REVERB_SCHEME=https
```

  
Then run these commands

```
php artisan config:cache
npm run build
```

  

### Now to update the apache configuration file to allow the WebSocket

#### First enable apache proxy module using these commands
```
sudo a2enmod proxy
sudo a2enmod proxy_wstunnel
sudo a2enmod proxy_http
```
This is my site config
```
<VirtualHost *:80>
	ServerName yourdomain.com
	ServerAdmin example@gmail.com

	DocumentRoot /srv/your-path/public

	<Directory /srv/your-path/public/>
		Options Indexes FollowSymLinks MultiViews
		AllowOverride All
		Order allow,deny
		allow from all
		Require all granted
	</Directory>

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

	RewriteEngine on
	RewriteCond %{SERVER_NAME} =yourdomain.com
	RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]
</VirtualHost>

  

<IfModule mod_ssl.c>
<VirtualHost *:443>
	ServerName yourdomain.com
	ServerAdmin example@gmail.com
  
	DocumentRoot /srv/your-path/public

	<Directory /srv/your-path/public/>
	Options Indexes FollowSymLinks MultiViews
	AllowOverride All
	Order allow,deny
	allow from all
	Require all granted
	</Directory>

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined
	SSLCertificateFile /etc/letsencrypt/live/yourdomain.com/fullchain.pem
	SSLCertificateKeyFile /etc/letsencrypt/live/yourdomain.com/privkey.pem
	Include /etc/letsencrypt/options-ssl-apache.conf

	# WebSocket proxying
	ProxyPreserveHost On
	ProxyPass /app/ ws://0.0.0.0:8080/app/
	ProxyPassReverse /app/ ws://0.0.0.0:8080/app/
</VirtualHost>
</IfModule>

```

### Replace

- `yourdomain.com` with your domain

- `/srv/your-path/public` with your path

- `example@gmail.com` with your email (optional)

Or you can just add this code inside the `<VirtualHost *:443>` tag
```
ProxyPreserveHost On
ProxyPass /app/ ws://0.0.0.0:8080/app/
ProxyPassReverse /app/ ws://0.0.0.0:8080/app/
```

Then run this command to load the new configuration
`sudo systemctl reload apache2`

### And remember don't forget to start reverb
```
php artisan reverb:start
```

And you are good to go.
Hope this helps

**Comment by gusberinger** [comment]
For Apache, putting this in my Pre Virtual Hosts config was sufficient:
```
<Location "/app/">
    ProxyPreserveHost On
    ProxyPass "http://127.0.0.1:8080/app/" upgrade=websocket
    ProxyPassReverse "http://127.0.0.1:8080/app/"

    RequestHeader set Host %{HTTP_HOST}s
    RequestHeader set Scheme %{REQUEST_SCHEME}s
    RequestHeader set SERVER_PORT %{SERVER_PORT}s
    RequestHeader set REMOTE_ADDR %{REMOTE_ADDR}s
    RequestHeader set X-Forwarded-For %{X-Forwarded-For}s
</Location>

<Location "/apps/">
    ProxyPreserveHost On
    ProxyPass "http://127.0.0.1:8080/app/" upgrade=websocket
    ProxyPassReverse "http://127.0.0.1:8080/app/"

    RequestHeader set Host %{HTTP_HOST}s
    RequestHeader set Scheme %{REQUEST_SCHEME}s
    RequestHeader set SERVER_PORT %{SERVER_PORT}s
    RequestHeader set REMOTE_ADDR %{REMOTE_ADDR}s
    RequestHeader set X-Forwarded-For %{X-Forwarded-For}s
</Location>
```

It does require the `mod_proxy_wstunnel` module.

**Comment by RafhaanVilla** [comment]
is there any way to use anything else than /app

    location /app {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://127.0.0.1:9000; 
      
    }
    
    in my application the user goes to /app to connect to the application, and its something that is difficult to change.

**Comment by CONSTANTlNE** [comment]
I  had same issue and struggled for several hours trying different configurations .... 

on local environment  i had  " implements ShouldBroadcast "  on event and then queue running ...  everything was working 

when i deploy the project on VPS , i thought everything would be the same ( I was ok with slight delay because of queue)  , but reverb was not working . laravel was not broadcasting the event ....  

Then by chance i switched from ShouldBroadcast   to ShouldBroadcastNow ....    and "magically" everything worked as it supposed to work ...   

production env   

( I had to use different port than 8080  because i had some other process there.)

  BROADCAST_CONNECTION=reverb
  
  REVERB_APP_ID=5555555
  REVERB_APP_KEY=SomeKeyHere
  REVERB_APP_SECRET=SuperSecret
  
  
  REVERB_HOST=yourdomain.com
  REVERB_SERVER_HOST=0.0.0.0
  REVERB_PORT=443
  REVERB_SERVER_PORT=6001
  REVERB_SCHEME=https
  
  VITE_REVERB_APP_KEY="${REVERB_APP_KEY}"
  VITE_REVERB_HOST="${REVERB_HOST}"
  VITE_REVERB_PORT="${REVERB_PORT}"
  VITE_REVERB_SCHEME="${REVERB_SCHEME}"


added snippet to Nginx : 


    location /app {
        proxy_http_version 1.1;
        proxy_set_header Host $http_host;
        proxy_set_header Scheme $scheme;
        proxy_set_header SERVER_PORT $server_port;
        proxy_set_header REMOTE_ADDR $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
 
        proxy_pass http://0.0.0.0:6001;
    }
